### PR TITLE
fix: validate gst_transporter_id

### LINKED ISSUE #1574: validate transporter id
https://docs.ewaybillgst.gov.in/Documents/NIC_TRANSIN_Advisory.pdf

### PR DESCRIPTION
**Pull request by vishakhdesai**
Closes #1574

image
image

### CODE REVIEW

File: india_compliance/gst_india/client_scripts/e_waybill_actions.js [hunk 1]
```diff
@@ -1121,6 +1127,12 @@ async function update_gst_tranporter_id(dialog) {
     dialog.set_value("gst_transporter_id", response.gst_transporter_id);
 }
 
+async function validate_gst_transporter_id(dialog) {
```
Comment (ljain112) [inline india_compliance/gst_india/client_scripts/e_waybill_actions.js]: validate for length before making request, also validate in backend

File: india_compliance/gst_india/doctype/gstin/gstin.py [hunk 2]
```diff
@@ -237,3 +244,34 @@ def get_formatted_response(response):
...
+
+    if frappe.db.exists("GSTIN", {"gstin": transporter_id, "is_transporter_id": 1}):
+        return frappe.get_doc("GSTIN", transporter_id)
+
+    company_gstin = get_company_gstin()
+    if not company_gstin:
+        return
+
+    try:
+        response = EWaybillAPI(company_gstin=company_gstin).get_transporter_details(
```
Comment (ljain112) [inline india_compliance/gst_india/doctype/gstin/gstin.py]: do we need to use archive info?

File: india_compliance/gst_india/doctype/gstin/gstin.py [hunk 3]
```diff
@@ -237,3 +244,34 @@ def get_formatted_response(response):
             "status": response.sts,
         }
     )
+
+
+def validate_transporter_id(transporter_id):
+    if not frappe.get_cached_value("GST Settings", None, "validate_gstin_status"):
+        return
+
+    if frappe.db.exists("GSTIN", {"gstin": transporter_id, "is_transporter_id": 1}):
```
Comment (ljain112) [inline india_compliance/gst_india/doctype/gstin/gstin.py]: duplicate code

File: india_compliance/public/js/utils.js [hunk 4]
```diff
@@ -52,7 +52,7 @@ Object.assign(india_compliance, {
         return in_list(frappe.boot.sales_doctypes, doctype) ? "Customer" : "Supplier";
     },
 
-    async set_gstin_status(field, transaction_date, force_update = 0) {
+    async set_gstin_status(field, transaction_date=null, force_update = 0) {
```
Comment (vorasmit) [inline india_compliance/public/js/utils.js]: not reqd IMO

File: india_compliance/gst_india/doctype/gstin/gstin.py [hunk 5]
```diff
@@ -84,7 +85,11 @@ def create_or_update_gstin_status(
     callback=None,
 ):
     doctype = "GSTIN"
-    response = _get_gstin_info(gstin=gstin, response=response)
+
+    if gstin[:2] == "88":
```
Comment (vorasmit) [inline india_compliance/gst_india/doctype/gstin/gstin.py]: Let's send the transaction date from the doc.

File: india_compliance/gst_india/doctype/gstin/gstin.py [hunk 6]
```diff
@@ -237,3 +250,26 @@ def get_formatted_response(response):
             "status": response.sts,
         }
     )
+
+
+def validate_transporter_id(transporter_id):
+    if not frappe.get_cached_value("GST Settings", None, "validate_gstin_status"):
```
Comment (vorasmit) [inline india_compliance/gst_india/doctype/gstin/gstin.py]: Instead how about checking for e-Waybill APIs

**Review by vorasmit — APPROVED** [review]

**Pull request merged**